src/linker/Linker.Steps/DynamicDependencyLookupStep.cs

```diff
@@ -0,0 +1,100 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using Mono.Cecil;
+using System.Diagnostics.CodeAnalysis;
+
+#nullable enable
+
+namespace Mono.Linker.Steps
+{
+	public class DynamicDependencyLookupStep : LoadReferencesStep
+	{
+		protected override void ProcessAssembly (AssemblyDefinition assembly)
+		{
+			var module = assembly.MainModule;
+
+			foreach (var type in module.Types) {
+				ProcessType (type);
+			}
+		}
+
+		void ProcessType (TypeDefinition type)
+		{
+			if (type.HasMethods) {
+				foreach (var method in type.GetMethods ()) {
+					var methodDefinition = method.Resolve ();
+					if (methodDefinition?.HasCustomAttributes != true)
+						continue;
+
+					ProcessDynamicDependencyAttributes (methodDefinition);
+				}
+			}
+
+			if (type.HasFields) {
+				foreach (var field in type.Fields) {
+					var fieldDefinition = field.Resolve ();
+					if (fieldDefinition?.HasCustomAttributes != true)
+						continue;
+
+					ProcessDynamicDependencyAttributes (fieldDefinition);
+				}
+			}
+
+			if (type.HasNestedTypes) {
+				foreach (var nestedType in type.NestedTypes) {
+					ProcessType (nestedType);
+				}
+			}
+		}
+
+		void ProcessDynamicDependencyAttributes (IMemberDefinition member)
+		{
+			Debug.Assert (member is MethodDefinition || member is FieldDefinition);
+
+			foreach (var ca in member.CustomAttributes) {
+				if (!IsPreserveDependencyAttribute (ca.AttributeType))
+					continue;
+#if FEATURE_ILLINK
+				Context.LogMessage (MessageContainer.CreateWarningMessage ($"Unsupported PreserveDependencyAttribute on '{member}'. Use DynamicDependencyAttribute instead.", 2029));
+#else
+				if (ca.ConstructorArguments.Count != 3)
+					continue;
+
+				if (!(ca.ConstructorArguments[2].Value is string assemblyName))
+					continue;
+
+				var assembly = Context.Resolve (new AssemblyNameReference (assemblyName, new Version ()));
+				if (assembly == null)
+					continue;
+				ProcessReferences (assembly);
+#endif
+			}
+
+			var dynamicDependencies = Context.Annotations.GetLinkerAttributes<DynamicDependency> (member);
```

Comment:
I don't think you should use `GetLinkerAttributes` here, you already iterate all attributes above so this is doing the same loop and also building cache for member which might not be reached at all.

Reply:
I [originally](https://github.com/mono/linker/pull/1215/commits/a1e1c3249a90c37ece24be0fb3329ebe505fb1e6#diff-8bcf8e9be10f0ae524d90bad71f67db3R81) optimized this a little more, but removed it based on this discussion: https://github.com/mono/linker/pull/1215#discussion_r431010530. If you think it is worth it, I could add back the check for this callsite, and only call `GetLinkerAttributes` if the previous loop found any `DynamicDependencyAttribute`s.